File: GolangSolutions/String/M2306_公司命名.go
/*
	@author Hyperspace
	@date 2024/09/25
	@file M2306_公司命名.go
*/

package String

func distinctNames(ideas []string) (ans int64) {
	st := [26]map[string]bool{}
	for i := range st {
		st[i] = map[string]bool{}
	}
	for _, s := range ideas {
		st[s[0]-'a'][s[1:]] = true
	}
	for i, a := range st {
		for _, b := range st[:i] {
			cnt := 0
			for s := range a {
				if b[s] {
					cnt++
				}
			}
			ans += int64(len(a)-cnt) * int64(len(b)-cnt)
		}
	}
	ans <<= 1

	return
}

File: CPPSolutions/LeetCode/String/M2306_公司命名.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long LL;

/**
 * @author Hyperspace
 * @date 2024/09/25
 * @file M2306_公司命名.cpp
 */
class Solution {
public:
    LL distinctNames(vector<string> &ideas) {
        unordered_set<string> st[26];
        for (auto &s: ideas) st[s[0] - 'a'].insert(s.substr(1));
        LL ans = 0;
        for (int i = 1, cnt; i < 26; i++) {
            for (int j = 0; j < i; j++) {
                cnt = 0;
                for (auto &s: st[i])
                    if (st[j].find(s) != st[j].end())
                        cnt++;
                ans += (st[i].size() - cnt) * (st[j].size() - cnt);
            }
        }

        return ans << 1;
    }
};